File: Python/107.Binary Tree Level Order Traversal II.py
#逆向分层返回一个二叉树所有值
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def levelOrder(self, root):
        self.result = []
        self.inorder(root,0)
        return self.result[::-1]

    def inorder(self,p,level):
        if p!=None:
            if len(self.result)<=level:
                self.result.append([p.val])
            else:
                self.result[level].append(p.val)
            self.inorder(p.left,level+1) if p.left!=None else None
            self.inorder(p.right,level+1) if p.right!=None else None



t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(2)
t4 = TreeNode(3)
t5 = TreeNode(3)
t6 = TreeNode(4)
t7 = TreeNode(4)
t1.left = t2
t1.right = t3
t2.left = t4
t2.right = t6
t3.right = t5
t3.left = t7


s = Solution()
print(s.levelOrder(t1))

File: C++/Tree/107.Binary Tree Level Order Traversal II.cpp
/*
从下到上，分层返回二叉树结点。
*/
/*
思路：每当有新的一层时，在结果数组中头部插入。然后在每层插入新的数时，下标为result.size()-level。
 */
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
	vector<vector<int>> result;
public:
	vector<vector<int>> levelOrderBottom(TreeNode* root) {
        if(root==NULL)	return result;
        helpLevelOrder(root,1);
        return result;
    }
    void helpLevelOrder(TreeNode* cur,int level)
    {
    	if(cur!=NULL)
    	{
    		if(result.size()<level)	result.emplace(result.begin(),vector<int> (1,cur->val));
    		else result[result.size()-level].push_back(cur->val);
    		helpLevelOrder(cur->left,level+1);
    		helpLevelOrder(cur->right,level+1);
    	}
    }
};
int main()
{
	Solution s;
	TreeNode t1(1),t2(2),t3(3);
	t1.right = &t2;
	t2.left = &t3;
	vector<vector<int>> out = s.levelOrderBottom(&t1);
	for(auto vec:out)
	{
		for(auto n:vec)
			cout<<n;
		cout<<endl;
	}
}

